File: Iteration 3/SystemController.py
import os
from JsonMethods import JsonMethods
from Student import Student
from Advisor import Advisor
from DepartmentScheduler import DepartmentScheduler
from DepartmentHead import DepartmentHead
from CourseRegistrationSystem import CourseRegistrationSystem
import logging

# Configure logging
logging.basicConfig(
    filename="error_logs.txt",
    level=logging.ERROR,
    format="%(asctime)s - %(levelname)s - %(message)s"
)

class LoginSystem:
    def __init__(self):
        self.json_methods = JsonMethods()
        self.students = []
        self.advisors = []
        self.staff_members = []
        self.courses = []
        self.load_data()

    def load_data(self):
        try:
            self.students = self.json_methods.load_all_students()
            self.advisors = self.json_methods.load_all_advisors()

            self.staff_members = []
            print("Loading Department Heads...")
            self.staff_members.extend(DepartmentHead.load_depthead())
            print("Loading Department Schedulers...")
            self.staff_members.extend(DepartmentScheduler.load_deptscheduler())

            self.courses = self.json_methods.load_course_json()

            print(f"{len(self.staff_members)} staff members loaded successfully.")
            print("All data loaded successfully.")
        except Exception as e:
            logging.error("Error loading data", exc_info=True)  # Log the error
            print(f"Error loading data: {e}")

    def start_login(self):
        """
        Starts the login process and directs to the appropriate menu based on the user role.
        """
        while True:
            username = input("Enter your username: ").strip()
            password = input("Enter your password: ").strip()

            # Student Login
            if username.startswith("o"):
                for student in self.students:
                    if student.get_username() == username and student.get_password() == password:
                        print(f"Welcome, {student.get_name()} {student.get_surname()}! (Student)")
                        self.student_menu(student)
                        return
                print("Invalid student username or password. Please try again.")

            # Advisor Login
            elif username.startswith("advisor"):
                for advisor in self.advisors:
                    if advisor.get_username() == username and advisor.get_password() == password:
                        print(f"Welcome, {advisor.get_name()} {advisor.get_surname()}! (Advisor)")
                        self.advisor_menu(advisor)
                        return
                print("Invalid advisor username or password. Please try again.")

            # Department Head Login
            elif username == "deptHead":
                for staff in self.staff_members:
                    if isinstance(staff, DepartmentHead) and staff.get_username() == username and staff.get_password() == password:
                        print(f"Welcome, {staff.get_name()} {staff.get_surname()}! (Department Head)")
                        self.department_head_menu(staff)
                        return
                print("Invalid department head username or password. Please try again.")

            # Department Scheduler Login
            elif username == "deptScheduler":
                for staff in self.staff_members:
                    if isinstance(staff, DepartmentScheduler) and staff.get_username() == username and staff.get_password() == password:
                        print(f"Welcome, {staff.get_name()} {staff.get_surname()}! (Department Scheduler)")
                        self.department_scheduler_menu(staff)
                        return
                print("Invalid department scheduler username or password. Please try again.")

            else:
                print("Invalid username. Please try again.")

    def student_menu(self, student):
        crs = CourseRegistrationSystem(self.json_methods, student, self.courses)

        while True:
            student.display_menu()
            student_choice = input("Enter your choice (or 'q' to quit): ").strip()

            if student_choice.lower() == 'b':
                break
            if student_choice.lower() == 'q':
                print("Exiting the program...")
                exit()

            if student_choice == "1":
                available_courses = crs.list_available_courses(student)
                for course in available_courses:
                    print(f"{course.get_course_id()} - {course.get_course_name()}")
            elif student_choice == "2":
                course_id = input("Enter the Course ID to request: ").strip()
                course = next((c for c in self.courses if c.get_course_id() == course_id), None)

                if not course:
                    print("Invalid Course ID. Please try again.")
                    continue

                crs.request_in_course(course, student)
            elif student_choice == "3":
                student.view_notifications(self.json_methods)
            elif student_choice == "4":
                student.display_schedule(self.json_methods)
            else:
                print("Invalid choice. Please try again.")

    def advisor_menu(self, advisor):
        advisor.refresh_advised_students()
        requests_map = {}

        while True:
            advisor.refresh_advised_students()
            requests_map = {}  # Reset requests_map each time
            advisor.display_menu()
            advisor_choice = input("Enter your choice (or 'q' to quit): ").strip()

            if advisor_choice.lower() == 'b':
                break
            if advisor_choice.lower() == 'q':
                print("Exiting the program...")
                exit()

            if advisor_choice == "1":
                requests_map = advisor.view_requests()
            elif advisor_choice == "2":
                requests_map = advisor.view_requests()
                if not requests_map:
                    print("No requests available to approve. Please try again.")
                    continue
                try:
                    request_index = int(input("Enter the request number to approve: ").strip()) - 1
                    advisor.approve_request_by_index(requests_map, request_index)
                except (ValueError, IndexError):
                    print("Invalid request number. Please try again.")
                    logging.error("Invalid request number. Please try again.", exc_info=True)   
            elif advisor_choice == "3":
                requests_map = advisor.view_requests()
                if not requests_map:
                    print("No requests available to reject. Please try again.")
                    continue
                try:
                    request_index = int(input("Enter the request number to reject: ").strip()) - 1
                    advisor.reject_request_by_index(requests_map, request_index)
                except (ValueError, IndexError):
                    print("Invalid request number. Please try again.")
                    logging.error("Invalid request number. Please try again.", exc_info=True)   
            else:
                print("Invalid choice. Please try again.")

    def department_head_menu(self, department_head):
        while True:
            department_head.display_menu()
            head_choice = input("Enter your choice (or 'q' to quit): ").strip()

            if head_choice.lower() == 'b':
                break
            if head_choice.lower() == 'q':
                print("Exiting the program...")
                exit()

            if head_choice == "1":
                department_head.add_course()
            elif head_choice == "2":
                department_head.update_course()
            elif head_choice == "3":
                department_head.display_all_courses()
            elif head_choice == "4":
                print("Logging out...")
                break
            else:
                print("Invalid choice. Please try again.")

    def department_scheduler_menu(self, scheduler):
        while True:
            scheduler.display_menu()
            scheduler_choice = input("Enter your choice (or 'q' to quit): ").strip()

            if scheduler_choice.lower() == 'b':
                break
            if scheduler_choice.lower() == 'q':
                print("Exiting the program...")
                exit()

            if scheduler_choice == "1":
                scheduler.print_all_courses()
            elif scheduler_choice == "2":
                scheduler.update_course_sections()
            elif scheduler_choice == "3":
                scheduler.reset_course_section()
            elif scheduler_choice == "4":
                scheduler.reset_all_course_sections()
            elif scheduler_choice == "5":
                print("Logging out...")
                break
            else:
                print("Invalid choice. Please try again.")

if __name__ == "__main__":
    login_system = LoginSystem()
    while True:
        login_system.start_login()

File: Iteration2/CSE3063F24P1_GRP14/src/SystemController.java
import java.io.IOException;
import java.util.*;

public class SystemController {
    public static void main(String[] args) throws IOException {
        JSONMethods jsonMethods = new JSONMethods();
        List<Course> courses = jsonMethods.loadAllCourses();
        List<Student> students = jsonMethods.loadAllStudents();
        List<Advisor> advisors = jsonMethods.loadAllAdvisors();

        // DepartmentScheduler'ı JSON'dan yükleme
        DepartmentScheduler departmentScheduler = jsonMethods.loadDepartmentScheduler();
        Scanner input = new Scanner(System.in);

        while (true) {
            String username;
            String password;
            String role = "";
            boolean isLoggedIn = false;
            Student loggedInStudent = null;
            Advisor loggedInAdvisor = null;

            //---------------Login-------------
            while (!isLoggedIn) {
                System.out.print("Username (write 'q' to quit): ");
                username = input.nextLine();

                if (username.equalsIgnoreCase("q")) {
                    System.out.println("Exiting program...");
                    return;
                }

                System.out.print("Password: ");
                password = input.nextLine();

                if (username.equals(departmentScheduler.getUsername()) && password.equals(departmentScheduler.getPassword())) {
                    role = "DepartmentScheduler";
                    System.out.println("Login successful, welcome Department Scheduler!");
                    isLoggedIn = true;
                    break;
                } else if (username.startsWith("o")) { // Student check
                    boolean found = false;
                    role = "Student";
                    for (Student student : students) {
                        if (student.getUsername().equals(username) && student.getPassword().equals(password)) {
                            found = true;
                            System.out.println("Login successful, welcome " + student.getName() + " (" + role + ")");
                            loggedInStudent = student;
                            isLoggedIn = true;
                            break;
                        }
                    }
                    if (!found) {
                        System.out.println("Wrong username or password.");
                    }
                } else if (username.startsWith("advisor")) { // Advisor check
                    boolean found = false;
                    role = "Advisor";
                    for (Advisor advisor : advisors) {
                        if (advisor.getUsername().equals(username) && advisor.getPassword().equals(password)) {
                            found = true;
                            System.out.println("Login successful, welcome " + advisor.getName() + " (" + role + ")");
                            loggedInAdvisor = advisor;
                            isLoggedIn = true;
                            break;
                        }
                    }
                    if (!found) {
                        System.out.println("Wrong username or password.");
                    }
                } else {
                    System.out.println("Wrong username or password.");
                }
            }

            // -----------------------Department Scheduler menu-----------------------------
            if (role.equals("DepartmentScheduler")) {
                while (true) {
                    departmentScheduler.getMenu();
                    String choiceInput = input.nextLine();

                    if (choiceInput.equalsIgnoreCase("q")) {
                        isLoggedIn = false;
                        break;
                    }

                    int choice;
                    try {
                        choice = Integer.parseInt(choiceInput);
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid input. Please enter a number or 'q' to go back.");
                        continue;
                    }

                    switch (choice) {
                        case 1: // View All Courses
                            departmentScheduler.printAllCourses();
                            break;

                        case 2: // Update Course Sections
                            System.out.println("Enter the course ID to update sections:");
                            String courseId = input.nextLine();
                            boolean success = departmentScheduler.updateCourseSections(courseId);
                            if (!success) {
                                System.out.println("Failed to update sections. Please try again.");
                            }
                            break;

                        case 3: // Reset All Course Sections
                            System.out.print("Are you sure you want to reset all course sections? (yes/no): ");
                            String confirmation = input.nextLine();
                            if (confirmation.equalsIgnoreCase("yes")) {
                                departmentScheduler.resetAllCourseSections();
                                System.out.println("All course sections have been reset.");
                            } else {
                                System.out.println("Reset operation canceled.");
                            }
                            break;

                        case 4: // Logout
                            System.out.println("Logging out...");
                            isLoggedIn = false;
                            break;

                        default:
                            System.out.println("Invalid choice. Please try again.");
                    }

                    if (!isLoggedIn) break;
                }
            }

            // -----------------------Advisor menu-----------------------------
            if (role.equals("Advisor")) {
                CourseRegistrationSystem crs = new CourseRegistrationSystem(null, courses);

                System.out.println("Advised Students for Advisor: " + loggedInAdvisor.getName());
                for (Student student : loggedInAdvisor.getAdvisedStudents()) {
                    System.out.println("Student ID: " + student.getStudentID() + " - Name: " + student.getName());
                    System.out.println("Requested Courses:");
                    for (Course course : student.getRequestedCourses()) {
                        System.out.println("  - " + course.getCourseId() + ": " + course.getCourseName());
                    }
                }



                while (true) {
                    loggedInAdvisor.getMenu();
                    String choiceInput = input.nextLine();

                    if (choiceInput.equalsIgnoreCase("q")) {
                        isLoggedIn = false;
                        break;
                    }

                    int choice;
                    try {
                        choice = Integer.parseInt(choiceInput);
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid input. Please enter a number or 'q' to go back.");
                        continue;
                    }

                    switch (choice) {
                        case 1: // See Requests
                            // JSON'dan öğrencileri güncelle
                            List<Student> updatedStudents = jsonMethods.loadAllStudents();

                            for (Student updatedStudent : updatedStudents) {
                                for (Student advisedStudent : loggedInAdvisor.getAdvisedStudents()) {
                                    if (advisedStudent.getStudentID().equals(updatedStudent.getStudentID())) {
                                        // Her kurs için mevcut listedeki tekrarları kontrol et
                                        for (Course updatedCourse : updatedStudent.getRequestedCourses()) {
                                            boolean exists = false;
                                            for (Course existingCourse : advisedStudent.getRequestedCourses()) {
                                                if (existingCourse.getCourseId().equals(updatedCourse.getCourseId())) {
                                                    exists = true;
                                                    break;
                                                }
                                            }
                                            // Eğer mevcut değilse ekle
                                            if (!exists) {
                                                advisedStudent.getRequestedCourses().add(updatedCourse);
                                            }
                                        }
                                    }
                                }
                            }

                            // Talepleri tablo olarak göster
                            System.out.println("Your advised student list:");
                            boolean hasRequests = false;

                            StringBuilder tableBuilder = new StringBuilder();
                            tableBuilder.append(String.format("%-5s %-20s %-20s %-10s %-40s\n", "No", "Student Name", "Surname", "Course ID", "Course Name"));
                            tableBuilder.append("---------------------------------------------------------------------------------------------\n");

                            int requestNo = 1;
                            for (Student student : loggedInAdvisor.getAdvisedStudents()) {
                                for (Course course : student.getRequestedCourses()) {
                                    hasRequests = true;
                                    tableBuilder.append(String.format("%-5d %-20s %-20s %-10s %-40s\n",
                                            requestNo++, student.getName(), student.getSurname(),
                                            course.getCourseId(), course.getCourseName()));
                                }
                            }

                            if (hasRequests) {
                                System.out.println(tableBuilder.toString());
                            } else {
                                System.out.println("There are no course requests at the moment.");
                            }
                            break;



                        case 2: // Approve Request
                            System.out.print("Enter the request number to approve: ");
                            int approveRequestIndex = input.nextInt();
                            input.nextLine(); // Consume newline

                            int count = 1;
                            Student selectedStudent = null;
                            Course selectedCourse = null;

                            // Yalnızca advisor'ın öğrencilerinin isteklerini dolaş
                            outerLoop:
                            for (Student student : loggedInAdvisor.getAdvisedStudents()) {
                                for (Course course : student.getRequestedCourses()) {
                                    if (count == approveRequestIndex) {
                                        selectedStudent = student;
                                        selectedCourse = course;
                                        break outerLoop;
                                    }
                                    count++;
                                }
                            }

                            if (selectedStudent != null && selectedCourse != null) {
                                loggedInAdvisor.approveRequestedCourse(selectedStudent, selectedCourse);
                            } else {
                                System.out.println("Invalid request number.");
                            }
                            break;


                        case 3: // Reject Request
                            System.out.print("Enter the request number to reject: ");
                            int rejectRequestIndex = input.nextInt();
                            input.nextLine(); // Consume newline

                            count = 1;
                            selectedStudent = null;
                            selectedCourse = null;

                            // Yalnızca advisor'ın öğrencilerinin isteklerini dolaş
                            outerLoop:
                            for (Student student : loggedInAdvisor.getAdvisedStudents()) {
                                for (Course course : student.getRequestedCourses()) {
                                    if (count == rejectRequestIndex) {
                                        selectedStudent = student;
                                        selectedCourse = course;
                                        break outerLoop;
                                    }
                                    count++;
                                }
                            }

                            if (selectedStudent != null && selectedCourse != null) {
                                CourseRegistrationSystem crs1 = new CourseRegistrationSystem(selectedStudent,courses);
                                loggedInAdvisor.rejectRequestedCourse(students,crs1,selectedStudent, selectedCourse);
                                System.out.println("Request rejected successfully.");
                            } else {
                                System.out.println("Invalid request number.");
                            }
                            break;

                        case 4: // Logout
                            System.out.println("Logging out...");
                            isLoggedIn = false;
                            break;

                        default:
                            System.out.println("Invalid choice. Please try again.");
                    }

                    if (!isLoggedIn) break;
                }
            }


            // -----------------------Student menu-----------------------------
            if (role.equals("Student")) {
                Student refreshedStudent = jsonMethods.loadStudent(loggedInStudent.getStudentID());
                if (refreshedStudent != null) {
                    loggedInStudent = refreshedStudent; // Güncel veriyi kullan
                }
                CourseRegistrationSystem crs = new CourseRegistrationSystem(loggedInStudent, courses);

                while (true) {
                    loggedInStudent.getMenu();
                    String choiceInput = input.nextLine();

                    if (choiceInput.equalsIgnoreCase("q")) {
                        isLoggedIn = false;
                        break;
                    }

                    int choice;
                    try {
                        choice = Integer.parseInt(choiceInput);
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid input. Please enter a number or 'q' to go back.");
                        continue;
                    }

                    switch (choice) {
                        case 1: // View Transcript
                            System.out.println("Your transcript:");
                            System.out.println(loggedInStudent.getTranscript().toString());
                            break;

                        case 2: // Request Course
                            refreshedStudent = jsonMethods.loadStudent(loggedInStudent.getStudentID());
                            if (refreshedStudent != null) {
                                loggedInStudent = refreshedStudent; // Güncel veriyi kullan
                            }
                            List<Course> availableCourses = crs.listAvailableCourses(loggedInStudent);

                            if (availableCourses.isEmpty()) {
                                System.out.println("No courses are available to request.");
                                break;
                            }

                            System.out.println("Available courses:");
                            System.out.println(crs.availableCoursesToString(availableCourses));

                            String courseCode;
                            Course selectedCourse = null;
                            boolean validCourseCode = false;

                            while (!validCourseCode) {
                                System.out.print("Enter the course ID you want to request (or 'q' to go back): ");
                                courseCode = input.nextLine();

                                if (courseCode.equalsIgnoreCase("q")) {
                                    break;
                                }

                                for (Course course : availableCourses) {
                                    if (course.getCourseId().equals(courseCode)) {
                                        selectedCourse = course;
                                        validCourseCode = true;
                                        break;
                                    }
                                }

                                if (!validCourseCode) {
                                    System.out.println("Invalid course ID. Please try again.");
                                }
                            }

                            if (selectedCourse != null) {
                                try {
                                    crs.requestInCourse(selectedCourse, loggedInStudent);
                                } catch (IOException e) {
                                    System.out.println("Error occurred while requesting the course: " + e.getMessage());
                                }
                            }
                            break;

                        case 3: // View Enrolled Courses
                            // JSON dosyasından en güncel öğrenciyi yükle
                            refreshedStudent = jsonMethods.loadStudent(loggedInStudent.getStudentID());
                            if (refreshedStudent != null) {
                                loggedInStudent = refreshedStudent; // Güncel veriyi kullan
                            }

                            System.out.println("Enrolled courses:");
                            List<Course> enrolledCourses = loggedInStudent.getEnrolledCourses();
                            if (enrolledCourses.isEmpty()) {
                                System.out.println("No enrolled courses.");
                            } else {
                                StringBuilder sb = new StringBuilder();
                                sb.append(String.format("%-10s %-40s\n", "Course ID", "Course Name"));
                                sb.append("------------------------------------------------------------\n");

                                for (Course course : enrolledCourses) {
                                    sb.append(String.format("%-10s %-40s\n",
                                            course.getCourseId(),
                                            course.getCourseName()));
                                }
                                System.out.println(sb.toString());
                            }
                            break;



                        case 4: // Display Schedule
                            System.out.println("Your schedule:");
                            loggedInStudent.displaySchedule(loggedInStudent);
                            break;

                        case 5: // Logout
                            System.out.println("Logging out...");
                            isLoggedIn = false;
                            break;

                        default:
                            System.out.println("Invalid choice. Please try again.");
                    }

                    if (!isLoggedIn) break;
                }
            }

        }
    }
}
